GSA-TTS/FAC
# Search Results Sorting & ALN Search Performance

# Search Results Sorting & ALN Search Performance

Epic: https://github.com/GSA-TTS/FAC/issues/2652

## Changes:
1. Sorting on the full result set:
    1. We no longer use the USWDS table sorting, we define the buttons manually.
    2. Clicking the sort button will use JS to resubmit the form with an `order_by` and `order_direction` field.
    3. Results will be sorted on the whole result set, not just the page you are seeing.
    4. Sorting will take you back to page one.
2. ALN search improvements:
    * Before, we would make a query to FederalAwards for every result that came up (thousands, in production).
    * Now, we make three total queries. One for results, a second for `finding_my_aln`, and a third for `finding_all_aln`. 

Small changes:
1. Search takes up the whole page. 
2. The search info box doesn't mention sorting on only one page anymore. 
3. A JS file was renamed.
4. The table header has a new subcomponent that includes the sort button and related info. 
5. State search has an accordion, and gets picked up by the reset button.

Changes also include some data loading things @jadudm was using to load enough results to get good tests going. 

## How to test:
1. Switch to this branch and run normally. You may need to run a collectstatic or re-up your docker containers. Navigate to search.
6. Ensure you have enough possible results to get several pages of results.
    * This can be done with data loading, and you can limit the number of results per page in `/dissemination/views.py`, line 76. 
7. Make a search. Attempt to sort. Verify that it reloads the page and that sorting has happened on the full result set, not just that page. 
8. Make an ALN search. Searching on `93` tends to give many results. Verify that you can sort by the new ALN columns, and that they also sort on the full result set. 

## Screenshots/Recordings:
The search screen size and sort buttons:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/1b269328-af40-48e5-b2ec-875a5c2f605b)


## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
